### PR TITLE
fit all concentrations at once

### PR DESCRIPTION
**Pull request by Farnazmdi**
Here is so far:
1. Wrote new functions in `stateCommon.py` for estimating the gamma parameters for all concentrations at once. 
2. Made a new class named `StateDistAll` in the `StateDistributionGamma.py` that accumulates the state distributions for all the concentrations. 
3. Made a new class named `AllObjs` in the `tHMM.py` to collect all the tHMM objects for different concentrations. 
4. In the  `M_E_step` in `BaumWelch.py` made an if statement that if the tHMMobj that is given is a list (which means it is for all concentrations at once) do a new method using those two newly made classes, otherwise does the old way of `M_E_step`.

Do you see a redundant step in these 4 that I have done @aarmey ? I'm actually a bit stuck and not sure if this is the right way. I will make a test to see if it works.

### CODE REVIEW

**Comment by aarmey** [comment]
I don't think we need a new state distribution; I think this can be accomplished within the structure of the current one. The emissions classes are a means for *storing* information, while we want the *behavior* to be different. I think the only two things we need are (1) a method that collects a list of gammas and observations across concentrations, and (2) a function that will take these lists and then perform the gamma distribution estimation. After the latter runs, the gamma distribution results can just be copied back into the current emissions distributions.

Maybe try writing #1 first? The second is essentially using the same approach to optimization of the distribution likelihoods, except while setting a [linear constraint](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.LinearConstraint.html#scipy.optimize.LinearConstraint) on the parameters between concentrations.

**Comment by Farnazmdi** [comment]
Will do it this way, thank you.

**Comment by Farnazmdi** [comment]
Sorry, just to clarify, for (1) you mean a method in the tHMM class, right? Because we won't access the statedistribution class anywhere after importing the lineages.

**Comment by aarmey** [comment]
It'd probably need to be outside of the tHMM class, but take a list of tHMM objects, since we have one tHMM per condition.

**Comment by Farnazmdi** [comment]
If we keep everything the same except that we gather the list of tHMM objects and for each of them from control to concentration_1, ... to concentration_last we pass the scale_estimated from the previous concentration to the current one that is being fitted, we don't need the linear_contraint and we don't change anything much. The conditions run separately but we control the upper bound of the scale parameter. How does this sound? Although, with this, we have to stick with the way we estimate shape_parameter.

**Comment by aarmey** [comment]
You have to fit them all at once because doing it step-wise makes the first condition special, which it's not.

**Comment by Farnazmdi** [comment]
I know I added many lines, but I can't figure out which part is not working... It is not being included in the fitting although I made sure the proper functions are being executed when we have more than 1 tHMMobj in the list that is being passed to the analyze_list. I really appreciate if you could take a quick look and see if something seems off.

File: lineage/BaumWelch.py
Comment (aarmey) [inline lineage/BaumWelch.py:155]: You should have this all work for one state at a time. Collecting everything across states makes it more complicated.

File: lineage/BaumWelch.py
Comment (aarmey) [inline lineage/BaumWelch.py:172]: I'd check that you're collecting all of the data together, and then assigning the results, correctly before worrying about the estimator. You can do this by simply skipping the actual estimator call and assigning the same parameter values as were there before.

File: lineage/states/StateDistributionGamma.py
Comment (aarmey) [inline lineage/states/StateDistributionGamma.py:193]: You know what the input and output to this should look like. You should write this on a separate branch, and then write a test that checks it works.

File: lineage/states/StateDistributionGamma.py
Comment (aarmey) [inline lineage/states/StateDistributionGamma.py:203]: Bernoulli shouldn't be handled in here. We could add constraints on it, too, but not right now. It can be handled by calling the normal estimators. Then, the constrained gamma solution can overwrite the normal gamma estimator.

File: lineage/states/stateCommon.py
Comment (aarmey) [inline lineage/states/stateCommon.py:87]: You know what the inputs and outputs to this should look like. This should be written with a separate branch and test.

**Review by aarmey — CHANGES_REQUESTED** [review]
Here are some thoughts in case it's not working. However, the real problem is manubot and something I need to fix. I'll make a separate PR.

**Comment by aarmey** [comment]
Ok, the manuscript build problem should now be fixed.

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py:153]: Do you mean that the functions should be written for one state or you mean I write a for loop over the state inside this function and do everything for each step at each iteration?

File: lineage/states/StateDistributionGamma.py
Comment (Farnazmdi) [reply lineage/states/StateDistributionGamma.py]: I moved this to another branch.

File: lineage/states/StateDistributionGamma.py
Comment (Farnazmdi) [reply lineage/states/StateDistributionGamma.py]: removed the bernoulli estimation from the estimator (that is in another branch)

File: lineage/states/stateCommon.py
Comment (Farnazmdi) [reply lineage/states/stateCommon.py]: Moved to another branch.

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py]: Should I collect all the cells and all the gammas when I am running the functions on each state? Or should I only pass those cells and those gammas for each specific state? I thought I should pass all and the gammas are there for distinguishing the states... I'm not sure I'm right

File: lineage/BaumWelch.py
Comment (aarmey) [reply lineage/BaumWelch.py:153]: I mean you should have a function that concerns itself with just one state. You can then call it N states numbers of times.

File: lineage/BaumWelch.py
Comment (aarmey) [reply lineage/BaumWelch.py]: Functions should only receive the data they need for the action at hand.

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py:153]: Thank you.

**Comment by Farnazmdi** [comment]
Here is in this branch:
1. A unit function in the `BaumWelch.py` that takes in all the cells and gammas for one state, and is supposed to run the estimator and assign the parameters.
2. A function that calls the unit function for each state and it called `do_M_E_step_atonce`.
3. A slight adjustment in the `do_M_step` and in the `fit_list()` in the tHMM, so when we are passing more than one concentration it runs the `...at_once` functions automatically.
I have the estimators in another branch.

File: lineage/figures/cartoons/gemcitabine.svg
Comment (aarmey) [inline lineage/figures/cartoons/gemcitabine.svg:3]: Did you intend to delete this?

File: lineage/figures/cartoons/lapatinib.svg
Comment (aarmey) [inline lineage/figures/cartoons/lapatinib.svg:3]: Did you intend to delete this?

File: lineage/tHMM.py
Comment (aarmey) [inline lineage/tHMM.py:189]: Don't you need to do this for `do_M_step` on line 197, too?

Comment (aarmey) [reply lineage/tHMM.py:189]: Honestly I don't remember why we have both `do_M_step` and `do_M_E_step`... I might be forgetting something critical but it'd be nice if they were combined or one just called the other after some setup.

File: lineage/BaumWelch.py
Comment (aarmey) [inline lineage/BaumWelch.py:169]: Why `vstack` one array?

**Review by aarmey — CHANGES_REQUESTED** [review]
A few things.

File: lineage/figures/cartoons/gemcitabine.svg
Comment (Farnazmdi) [reply lineage/figures/cartoons/gemcitabine.svg:3]: Yes, actually, these are generated every time we run the whole figures. I know I had written something in the makefile to delete them each time, but not sure why it is not doing that.

File: lineage/figures/cartoons/lapatinib.svg
Comment (Farnazmdi) [reply lineage/figures/cartoons/lapatinib.svg:3]: same as above.

File: lineage/tHMM.py
Comment (Farnazmdi) [reply lineage/tHMM.py:189]: Because the `do_M_step` takes the list of tHMM objects, I wrote this if statement inside it, but the `do_M_E_step` only applies on one tHMM object at a time, so I had to keep the if statement here. But you are right, there is no need to separate them. I will combine the two right away.

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py:169]: In this, I followed the pattern that was being done for each population, only here I have the data of concentrations appended in a list, so I looped over those inside the list. I don't think it is one array; it should be the matrix that is keeping the N x K gammas for each lineage and this should be stacking all the lineages of the same condition.

File: lineage/tHMM.py
Comment (Farnazmdi) [reply lineage/tHMM.py]: I removed the `do_M_E_step` but didn't removed the one for `atonce_...`.
Also, it seemed to me that the initialization in the `fit_list()` was not really doing anything and I removed it and built a few of the figures, they were not different at all.
Is it necessary to have random initialization before the first step of EM? I think because we initialize everything in the classes this may not be necessary...

Comment (aarmey) [reply lineage/tHMM.py]: This all sounds reasonable. Is this ready for review then?

Comment (Farnazmdi) [reply lineage/tHMM.py]: It turned out that by removing the initialization, apparently, it messed with calculating the `lowess` for figure S04. I searched the error (which was this: `FloatingPointError: invalid value encountered in true_divide`) and it doesn't seem to be easily fixable! So brought the `do_M_E_step` back so we could call only that in the initialization part in `fit_list`.

**Comment by Farnazmdi** [comment]
Sorry, I had forgotten to re-request your review on this.

**Review by aarmey — COMMENTED** [review]
This all looks good to me. Let's leave this here, and then we will add the other parts by making PR's onto this branch. Finally, this will get merged with master when all the parts are done.

**Comment by Farnazmdi** [comment]
Sure, sounds great. The estimator PR is close to ready, there are few issues that I need to figure out.

File: lineage/states/stateCommon.py
Comment (aarmey) [inline lineage/states/stateCommon.py:115]: @Farnazmdi I think what will help more than anything else is to pass the parameters from the last iteration in as a starting point (`x0`). Now, the first iteration won't satisfy the bounds, so you could write a check in here to override the starting point if it isn't valid. We do this with the other estimator, and it means that the first iteration takes the same amount of time, but subsequent ones are much faster.

Comment (Farnazmdi) [reply lineage/states/stateCommon.py:118]: Sorry, this had slipped out of my sight. I'm on it.

File: lineage/states/StateDistributionGamma.py
Comment (aarmey) [inline lineage/states/StateDistributionGamma.py:222]: @Farnazmdi why would running the function twice help to speed it up?

Comment (Farnazmdi) [reply lineage/states/StateDistributionGamma.py:222]: I know sorry, I am working on it in local, it should have been recursive, not running that twice. But the problem is that when I call the function inside it, it doesn't get out of it and just keeps running itself. I should find a good stopping point.

Comment (aarmey) [reply lineage/states/StateDistributionGamma.py:222]: Why would making it recursive help in any way?

Comment (aarmey) [reply lineage/states/StateDistributionGamma.py:222]: When I said "last iteration", I meant last iteration of Baum-Welch, not of the optimization here. Think through the problem. There is also the separated estimator as an example.

Comment (Farnazmdi) [reply lineage/states/StateDistributionGamma.py]: It should be fixed now. I appreciate if you could review it.

**Review by aarmey — COMMENTED** [review]
This looks right! Let me review it once more tomorrow morning before merging.

**Comment by Farnazmdi** [comment]
Thank you.

File: lineage/BaumWelch.py
Comment (aarmey) [inline lineage/BaumWelch.py:92]: Does this mean this will be applied in Fig. 10? Do the emissions we generate in that figure satisfy our constraint here? We'll also need to adjust the degrees of freedom provided to AIC (not here but make an issue).

File: lineage/BaumWelch.py
Comment (aarmey) [inline lineage/BaumWelch.py:157]: This is collecting the emissions, right? Given that these don't depend on the state, why not do this in `do_M_E_step_atonce` rather than `do_M_E_step_1state`?

File: lineage/states/stateCommon.py
Comment (aarmey) [inline lineage/states/stateCommon.py:145]: Is this needed? What was happening when the upper bound was just uniformly 100?

File: lineage/tHMM.py
Comment (aarmey) [inline lineage/tHMM.py:194]: Remove printing.

File: lineage/tHMM.py
Comment (aarmey) [inline lineage/tHMM.py:172]: Remove tolerance change.

File: lineage/states/stateCommon.py
Comment (aarmey) [inline lineage/states/stateCommon.py:147]:
```suggestion
    res = fmin_cobyla(func=negative_LL_atonce, x0=x0, cons=[constr, constr2, constr3], args=arrgs, consargs=(), maxfun=500000, rhoend=1e-6, rhobeg=2.0)
```

Remove printing, lower final tolerance, and allow slightly larger steps at the beginning.

**Review by aarmey — CHANGES_REQUESTED** [review]
Just some small changes, then this looks ready!

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py:92]: Yes, I hope it can help with solving the issue that we had with figure 10 (every time it looked different).
I will make the issue right now.

Comment (Farnazmdi) [reply lineage/BaumWelch.py:92]: I have to look into the emissions, but I think the data is the same, so should be the emissions.

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py:157]: I removed `do_M_E_step_1state` and all is included in the `do_M_E_step_atonce` now.

File: lineage/states/stateCommon.py
Comment (Farnazmdi) [reply lineage/states/stateCommon.py:145]: Maybe I'm wrong, but I expect the shape parameter to be larger than the scale parameter.
Given that `shape x scale` should indicate the average of phase lengths, I thought we should restrict them a bit more because their multiplication should not be a very large number.

File: lineage/BaumWelch.py
Comment (Farnazmdi) [reply lineage/BaumWelch.py:92]: I will take a closer look at this in another PR.

File: lineage/states/stateCommon.py
Comment (Farnazmdi) [reply lineage/states/stateCommon.py:145]: Should I bring 800 constraint back?

Comment (aarmey) [reply lineage/states/stateCommon.py:145]: I think just 100 upper bound across the board? Mainly I think it should be consistent with the bounds in the separate estimator.

Comment (Farnazmdi) [reply lineage/states/stateCommon.py:145]: Of course. Will do.

**Review by aarmey — APPROVED** [review]
Ready to merge if it all builds fine.

**Pull request merged**
